### ISSUE TITLE
trigger_t objects point to "last" event, which may have been freed or recycled.

### ISSUE THREAD

**Issue by edwardalee** [lf-lang/reactor-c#145]
The trigger_t struct has a field `last` that points to the last event scheduled by that trigger. That field is used to implement the drop policy to enforce minimum spacing.  Unfortunately, the field may point to an event that has been recycled (or worse, freed).

The field needs to be replaced with a `last_time` field initialized to `NEVER`. However, this also means that the pqueue function `pqueue_find_equal_same_priority` needs to be redone to take a trigger and time argument instead of an event.

**Comment by lhstrh** [comment]
Related to #215.

**Comment by SheaFixstars** [comment]
As mentioned in this issue this is related to the `last` trigger field. 

When that event is done it gets recycled but the trigger->last is not removed so when the next event goes to run it could trigger the drop policy due to the `last` event being reused, assuming it does not crash due to freed memory. This leads to lost events as high as 50% lost in testing that I see. This issue is also hit with both "replace" and "defer" policy as well. 

- If there is a min_space over 0 then the issue is triggered, regardless of the policy.
- Only triggers when there is no entry in the queue when a new entry comes in

The test below is using a logical action with a drop policy, the queued items will only ever have one item in the queue at a time so they don't overlap and properly trigger the drop policy.

The following program is setup as follows:
 - Timer runs every 500ms
 - Schedules logical action with 1ms delay, 2000ms min_space "drop" policy
   - Prints out the Action statement
   - Current is current logical time, scheduled is the expected logical runtime
 - Logical action prints out the run statement when run
   - Time and microstep are printed out, the Msg Logical is when it was expected to run

The issue can be seen in the following:
```
(Timer) Schedule - Current: 0, Scheduled: 1000000, Count:0
(Action) Run - Time:1000000 Microstep:0, Msg Logical:1000000, Count:0
(Timer) Schedule - Current: 500000000, Scheduled: 501000000, Count:1
(Timer) Schedule - Current: 1000000000, Scheduled: 1001000000, Count:2
(Action) Run - Time:1001000000 Microstep:0, Msg Logical:1001000000, Count:2
(Timer) Schedule - Current: 1500000000, Scheduled: 1501000000, Count:3
(Timer) Schedule - Current: 2000000000, Scheduled: 2001000000, Count:4
(Action) Run - Time:2001000000 Microstep:0, Msg Logical:2001000000, Count:4
(Timer) Schedule - Current: 2500000000, Scheduled: 2501000000, Count:5
(Timer) Schedule - Current: 3000000000, Scheduled: 3001000000, Count:6
(Action) Run - Time:3001000000 Microstep:0, Msg Logical:3001000000, Count:6
```
The correct output after the proposed fix is the following:
```
(Timer) Schedule - Current: 0, Scheduled: 1000000, Count:0
(Action) Run - Time:1000000 Microstep:0, Msg Logical:1000000, Count:0
(Timer) Schedule - Current: 500000000, Scheduled: 501000000, Count:1
(Action) Run - Time:501000000 Microstep:0, Msg Logical:501000000, Count:1
(Timer) Schedule - Current: 1000000000, Scheduled: 1001000000, Count:2
(Action) Run - Time:1001000000 Microstep:0, Msg Logical:1001000000, Count:2
(Timer) Schedule - Current: 1500000000, Scheduled: 1501000000, Count:3
(Action) Run - Time:1501000000 Microstep:0, Msg Logical:1501000000, Count:3
(Timer) Schedule - Current: 2000000000, Scheduled: 2001000000, Count:4
(Action) Run - Time:2001000000 Microstep:0, Msg Logical:2001000000, Count:4
(Timer) Schedule - Current: 2500000000, Scheduled: 2501000000, Count:5
(Action) Run - Time:2501000000 Microstep:0, Msg Logical:2501000000, Count:5
(Timer) Schedule - Current: 3000000000, Scheduled: 3001000000, Count:6
(Action) Run - Time:3001000000 Microstep:0, Msg Logical:3001000000, Count:6
```


I am proposing a fix of unsetting the `last` event within the event pop if the `last` matches the event being popped. I am fine to do a pull request for this but want to see peoples thoughts as the initial proposal is a larger rework. It may be worth bringing this in until the rework is done if the fix works fine to patch the issue temporarily.

Looking at adding the following code at https://github.com/lf-lang/reactor-c/blob/61261af0b93571226e1aa91ed314d1e3b13ce0b2/core/reactor_common.c#L426
```
//Clear the trigger last if it matches the current event
if (event->trigger->last == event)
{
   event->trigger->last = NULL;
}
```

Attached is the results for the issue and fixed version for the defer, replace, and drop policies.

[Policy_Test_Defer_Single_Queue_Item.lf.txt](https://github.com/lf-lang/reactor-c/files/13590396/Policy_Test_Defer_Single_Queue_Item.lf.txt)
[Policy_Defer_Results.txt](https://github.com/lf-lang/reactor-c/files/13590380/Policy_Defer_Results.txt)
[Policy_Defer_Results_Fixed.txt](https://github.com/lf-lang/reactor-c/files/13590381/Policy_Defer_Results_Fixed.txt)

[Policy_Test_Drop_Single_Queue_Item.lf.txt](https://github.com/lf-lang/reactor-c/files/13590397/Policy_Test_Drop_Single_Queue_Item.lf.txt)
[Policy_Drop_Results.txt](https://github.com/lf-lang/reactor-c/files/13590382/Policy_Drop_Results.txt)
[Policy_Drop_Results_Fixed.txt](https://github.com/lf-lang/reactor-c/files/13590383/Policy_Drop_Results_Fixed.txt)

[Policy_Test_Replace_Single_Queue_Item.lf.txt](https://github.com/lf-lang/reactor-c/files/13590398/Policy_Test_Replace_Single_Queue_Item.lf.txt)
[Policy_Replace_Results.txt](https://github.com/lf-lang/reactor-c/files/13590384/Policy_Replace_Results.txt)
[Policy_Replace_Results_Fixed.txt](https://github.com/lf-lang/reactor-c/files/13590385/Policy_Replace_Results_Fixed.txt)

**Comment by lhstrh** [comment]
Fixed in #318.
